### PR TITLE
#365: add support for Kokkos::Array

### LINKED ISSUE #365: Add support for serializing Kokkos::Array
Kokkos::Array is used as a standin for std::array` in device code, we should support serializing it

### PR DESCRIPTION
**Pull request by nmm0**
Closes #365

### CODE REVIEW

File: tests/unit/test_kokkos_serialize_array.cc [hunk 1]
```diff
@@ -0,0 +1,74 @@
...
+    auto& outArray = *deserialized;
+
+    for (size_t i = 0; i < N; ++i )
+        ASSERT_EQ(refArray[i], outArray[i]);
+}
+
+TEST_F(KokkosArrayTest, test_kokkos_array) {
+    using namespace ::checkpoint;
+
+    auto arr1 = Kokkos::Array{ 1, 2, 3, 4, 5 };
```
Comment (cz4rs) [inline tests/unit/test_kokkos_serialize_array.cc]: This is not supported in Kokkos `4.1.0` which is used in CI:
https://github.com/DARMA-tasking/magistrate/blob/3d60dd6268c6f2b37086212eed86866ac36f9a2d/ci/docker/ubuntu-22.04-clang-cpp.dockerfile#L57

Works fine with `4.3.01` for me.

Comment (nmm0) [reply tests/unit/test_kokkos_serialize_array.cc]: ah yeah I forgot CTAD was only added recently

File: src/checkpoint/container/kokkos_array.h [hunk 2]
```diff
@@ -0,0 +1,73 @@
...
+
+#include <Kokkos_Array.hpp>
+
+namespace checkpoint {
+
+#if KOKKOS_VERSION_LESS(4, 4, 0)
+template <typename Serializer, typename T, size_t N, class Proxy>
+void serialize(Serializer& s, Kokkos::Array<T, N, Proxy>& array) {
+  static_assert(std::is_void_v< Proxy >, "Magistrate does not support serializing Kokkos Arrays with proxies");
+  dispatch::serializeArray(s, &array[0], array.size());
```
Comment (PhilMiller) [inline src/checkpoint/container/kokkos_array.h]:
```suggestion
  dispatch::serializeArray(s, array.data(), array.size());
```
?

Comment (nmm0) [reply src/checkpoint/container/kokkos_array.h]: ah thanks, overlooked that

**Review by cz4rs — APPROVED** [review]
Looks good to me.

**Comment by cz4rs** [comment]
Running the `check_guards.sh` or `generate_header_guards_and_license.py` does not modify any files for me :thinking: Not sure why the build is failing in CI.

**Comment by cz4rs** [comment]
@nmm0 Got it, you just need to rebase on top of `develop`, re-run the script and a bunch of changes will show up:
```diff
 //                                kokkos_array.h
-//                 DARMA/checkpoint => Serialization Library
+//                 DARMA/magistrate => Serialization Library
 //
```

**Comment by nmm0** [comment]
ahh makes sense ill rebase

**Review by cz4rs — APPROVED** [review]

**Review by lifflander — APPROVED** [review]
Looks good.

**Review by nlslatt — APPROVED** [review]

**Pull request merged**